src/Client/HedgedConnectionsFactory.h

```diff
@@ -0,0 +1,132 @@
+#pragma once
+
+#if defined(OS_LINUX)
+
+#include <Common/TimerDescriptor.h>
+#include <Common/Epoll.h>
+#include <Common/FiberStack.h>
+#include <Common/Fiber.h>
+#include <Client/ConnectionEstablisher.h>
+#include <Client/ConnectionPoolWithFailover.h>
+#include <Core/Settings.h>
+#include <unordered_map>
+#include <memory>
+
+namespace DB
+{
+
+/** Class for establishing hedged connections with replicas.
+  * The process of establishing connection is divided on stages, on each stage if
+  * replica doesn't respond for a long time, we start establishing connection with
+  * the next replica, without cancelling working with previous one.
+  * It works with multiple replicas simultaneously without blocking by using epoll.
+  */
+class HedgedConnectionsFactory
+{
+public:
+    using ShuffledPool = ConnectionPoolWithFailover::Base::ShuffledPool;
+
+    enum class State
+    {
+        READY,
+        NOT_READY,
+        CANNOT_CHOOSE,
+    };
+
+    struct ReplicaStatus
+    {
+        ReplicaStatus(ConnectionEstablisher connection_stablisher_) : connection_establisher(std::move(connection_stablisher_))
+        {
+        }
+
+        ConnectionEstablisher connection_establisher;
+        TimerDescriptor change_replica_timeout;
```

Comment:
That's interesting that `connection_establisher` has it's own timeout descriptor.
But those descriptors are different.
`change_replica_timeout` is totals timeout to establish whole connection, but timeout inside `connection_establisher` is separate for each step.

I want such kind of comment here :)

Also, there is a thought that maybe we can put this timer inside `connection_establisher`, so it will be clear that `connection_establisher` has two different timers, and it is for reason. (I'm not  sure about it).

Reply:
I just thought that it will be strange to put change_replica_timeout inside ConnectionEstablisher, because this class is responsible for establishing connection and don't need to know about changing replicas logic. I will write a comment about it

Reply:
> this class is responsible for establishing connection and don't need to know about changing replicas logic

Yes, reasonable.